pkg/arvo/lib/lv.hoon

```diff
@@ -0,0 +1,546 @@
+::
+::::  Vector type in single-precision floating-point @rs
+  ::
+|%
+++  lvs
+  ^|
+  ~%  %lvs  ..is  ~
+  |_  r/$?($n $u $d $z)   :: round nearest, round up, round down, round to zero
+  ::
+  ::  Manipulators
+  ::
+  ::    Zeroes
+  ++  zeros
+    ~/  %zeros
+    |=  n=@ud  ^-  @lvs
+    ~_  leaf+"lagoon-fail"
+    :: There is a vanishingly small chance of getting a NaN in a vector which
+    :: collides with this, but it's like getting a hash table collision, very
+    :: very unlikely.  Also the Urbit NaN is hard-coded, `SINGNAN`.
+    =/  nan  0xffff.ffff
+    `@lvs`(not 5 +(n) (rap 5 ~[(fil 5 n nan) (not 5 1 n)]))
```

Comment:
Yes.  This sort of refactoring is what is needed so it doesn't feel so cobbled-together.

Pending the actual atom form, of course, for which we need more weigh-ins.

Reply:
This yields the desired behavior for the current atom shape.  I've added it locally (will pull them in as a block).  I'll use :rocket: to indicate integration on each of these.

Reply:
Yeah, was definitely aiming for "preserve semantics" here. Though more generally the atom format is "row-major with some trailer words", swapping in half-word concat for cantor or 0x1 for (lent) or w/e would be fairly straightforward - idk if anyone else has opinions on the atom form.

More food for thought: ideally `.1_0j1` would be the complex basis vector…

Reply:
I'm a little bugged that `@c` and `@i` are both occupied already for simple complex atoms...

Reply:
Yeah - otoh we could stretch the mnemonic of "real" somewhat and have `@rsr` `@rdr` etc be plain floats, and `@rs` be "LSB real then complex parts", since after all casting a 32-bit float to a pair of a 32-bit real and 32 trailing zero bits is perfectly valid.

Reply:
Then in my representation, use e.g. the top bit of the shape-length byte for "is complex", so `0x1` is the "vector of reals" marker and `0x81` is the "vector of complex" marker - means if you want transparent scalars you can't have a vector of 1 real, just a vector of 1 complex, which is kinda inefficient, but not sure it's a huge deal. 
(Alternately do something clever with NaNs and a non-word-aligned trailing byte)

Or you could give up on untagged scalars and full-on have both the shape and complexity and while we're at it bitwidth in the trailer, producing a new `@rt` "float tagged / float tensor" format. Would probably want additional literal syntax for "one float/complex, not in a vector or matrix" and generally make the tensor one less "just a float but plural" - is `..` a used prefix?

Reply:
In general this gets you quite an aura stack, but I think they all telescope properly:
```
@rt - float tagged
@rts - float tagged single-precision
@rtsu - float tagged single-precision unitary(singular)
@rtsur - float tagged single-precision one real
@rtsuc - float tagged single-precision one complex
@rtsvr - float tagged single-precision vector of real
@rtsmc - float tagged single-precision matrix of complex
(@rtr - float tagged some-precision of some-shape of reals, etc)
```

Reply:
#3759 up with `AjB` parsing/printing - haven't done any of the math support, kinda want to pull it out into zuse and leave hoon.hoon with just scalar mul/div/round/dragon4 as necessary for textual serdes.

Reply:
Following up:  this implies that the primitive complex value is actually a 1×1 complex matrix or a 1× complex vector.  Should there also be a "bare" complex aura?  Do we back into a corner by only having in-structure complexes?

Reply:
Let me write up a proposal with several possibilities and submit it as a separate issue.

Reply:
I've added my four cases to #3759 now for discussion.  Pending that, I'll start resolving the atom structure for `@l` or `@rt`.

Reply:
The main distinction in my mind is "are @rd bit-compatible with lagoon scalars", sort of in the same way that `@ux` can be a "view" onto `@t` but even more tightly corresponding.

If we go the route of "rank bitwidth etc live in the auras"(and only the auras), then `@lus` would be a "lagoon scalar", which is different from `@rs` only in that the trailing imaginary halfword can be nonzero. This gives you a cleaner separation between tensor/complex-capable and tensor-incapable code, with speedup from avoiding dynamic dispatch, at the cost of having to care about that this distinction in all calling code. (inserting various typecasts, duplicating interfaces for all combinations, doing the same if the calling code is itself a general-purpose library.)

If we go the route of "rank bitwidth etc are encoded in the value"(i.e. the trailing byte scheme), then generic `@rt` is usable everywhere, and the more specific auras are only relevant for constants / when you are sure you want only matrices or only reals or w/e. (Though as more of a "flagset" than conventional aura nesting, we'd either want only some of the tree expressly encoded, or to change compiler nesting semantics to be more filter-list like.) This gives you an easier to follow surface area, but requires runtime conversion to interact with existing float code ~~which there isn't any of~~, and of course if you *do* type everything as just `@rt`, you might end up with a complex vector where you wanted a real matrix. Library functions like `add` in particular will want to be superficially-wet, in that there is a core implementation that works on all `@rt`(but checks size and probably-format compatibility), and then a wrapper around it that calculates the result aura from that of `a` and `b`(and type-fails on statically known mismatch).

There is also a possibility of e.g. encoding the rank but not the individual float `bloq`, which I think was my earlier proposal, but it seems likely correct to go all-in one way or the other. And given the *shape* is not in the aura anyway, I'm leaning towards "~all\~ the runtime type information".